skeleton/cypress/aurelia_project/tasks/cypress.ext

```diff
@@ -7,14 +7,27 @@ import * as cypress from 'cypress';
 import * as config from '../../cypress.config';
 // @endif
 import { CLIOptions } from 'aurelia-cli';
+import { default as runApp } from './run';
 
-export default (cb) => {
+const runCypress = (cb) => {
   if (CLIOptions.hasFlag('run')) {
     cypress
       .run(config)
-      .then(results => (results.totalFailed === 0 ? cb() : cb('Run failed!')))
+      .then(results => {
+        (results.totalFailed === 0 ? cb() : cb('Run failed!'))
+        process.exit(0);
```

Comment:
It is at least incorrect for failed test. The process will exit with non-zero code if tests failed, I think that works before.
If you want to kill the web server, you need to find way to kill it. I think at least for cli-bundler, browsersync has an api to stop the server. webpack-dev-server probably got an api too.

Reply:
>  I think that works before.

When you run it in headless mode, and all tests are passed the process is still in running state, preventing the rest of the commands to be run, it is described on [discourse](https://discourse.aurelia.io/t/how-to-start-aurelia-site-before-cypress-testing/2375), but when it has failing tests, there is even no need to `process.exit(0)`

What do you think about this workaround, the listening server will be passed to the callback of the imported runApp function: 

In **run.ts** we could pass the server to the callback, done, method:
```ts

  server.listen(opts.port, opts.host, function(err) {
    if (err) throw err;

    // reportWebpackReadiness(opts);
    done(server);
  });
```
Then in the **cypress.ts** have something like this to **close** the server with **_server.close()_**: 
```ts
import { default as runApp } from './run';

const runCypress = (cb) => {
  if (CLIOptions.hasFlag('run')) {
    cypress
      .run(config)
      .then(results => {
        (results.totalFailed === 0 ? cb() : cb('Run failed!'))
      })
      .catch(cb);
  } else {
    cypress.open(config);
  }
}

export default (cb) => {
  if (CLIOptions.hasFlag('start')) {
    runApp((server) => {
      runCypress((msg) => {
        server.close();
        msg ? cb(msg) : cb();
      });
    });
    return;
  }

  runCypress(cb);
};
```

That being said, I think the following approach will be more general and it will respect the failing tests as well: 

```ts
cypress
      .run(config)
      .then(results => {
        if (results.totalFailed === 0) {
          cb();
          process.exit(0);
        } else {
          cb('Run failed!');
          process.exit(-1);
        }
      })
      .catch(cb);
```

I think the latter will have cleaner code, while the former might be more correct, ideas?

Reply:
I tested it with a `Dockerfile` for automation purposes, and when we have tests that fail, `Cypress` itself returns a non-zero code, and there is no need to interfere with `process.exit(n)`, so I guess the following will suffice: 

```ts
cypress
      .run(config)
      .then(results => {
        if (results.totalFailed === 0) {
          cb();
          process.exit(0);
        } else {
          cb('Run failed!');
        }
      })
      .catch(cb);
```

Reply:
I tested the idea of `done(server)`, that killed `au run`, because gulp think the argument is an error for `done()` callback.

But `done(undefined, server)` worked, probably too hacky?

Reply:
```js
function _runWebpack(done) {
  // ...
  done(server);
  // ...
}

function runWebpack(done) {
  return _runWebpack(() => done());
}

const run = gulp.series(
  configureEnvironment,
  runWebpack
);

export { run as default, _runWebpack };
```
This is probably better. Then e2e task uses _runWebpack directly.

Reply:
I suggested the mechanism of passing the server, but as you said it seems too hacky, and I'm afraid it becomes more complicated when it comes to other server types rather than webpack dev server; besides the sources are now coupled more and that could be harmful in a long run, in terms of maintainability.

 I would like to go with the following approach:

```ts
const runCypress = (cb) => {
  if (CLIOptions.hasFlag('run')) {
    cypress
      .run(config)
      .then(results => {
        if (results.totalFailed === 0) {
          cb();
          console.log( "Finished successfully" );
          process.exit(0);
        } else {
          console.log(`Number of failed tests: ${results.totalFailed}`);
          cb('Run failed!');
        }
      })
      .catch(cb);
  } else {
    cypress.open(config);
  }
}
```

That way, when anything rather than test fails the cypress will exit with the `catch` block and if the tests fail/pass it will go through the then clause, which either terminates the process automatically if all tests have been passed or goes through the running pipeline with failure.

These are the output on my machine in failure and passed states respectively: 

** Failed Tests**

![image](https://user-images.githubusercontent.com/4405518/61167755-539cc980-a558-11e9-92b1-b716cad9013d.png)

** All Tests Passed **

![image](https://user-images.githubusercontent.com/4405518/61167700-a6c24c80-a557-11e9-94c5-67ee39a983a0.png)

Reply:
When fail, probably need to double check whether `process.exit()` kills off dev server too (just try to visit http://localhost:5050 again after tests finished).

Did I miss something? I didn't see how dev server is terminated when tests passed.

The exit code 0 is not right for failure branch. See the comments for protractor.

Reply:
> When fail, probably need to double check whether process.exit() kills off dev server too (just try to visit http://localhost:5050 again after tests finished).

I didn't use `process.exit()` in the case of failure, it is the old way by just calling `cb('Run failed!');`

>Did I miss something? I didn't see how dev server is terminated when tests passed.

 Here, I used `process.exit(0);` and also called `cb()`, since it is a success path, `process.exit(0)` has no harm, right?


>The exit code 0 is not right for failure branch. See the comments for protractor.

I did not return any exit code! could you tell me where you've seen that? Is it the `uid` in the output:

```bash
{ uid: 0,
  name: '<anonymous>',
  branch: false,
  error: 'Run failed!',
  duration: [ 17, 755040600 ],
  time: 1563172848002
}
```


**PS:** I haven't pushed any changes yet, I postponed that to the time when this discussion resolved, I could push changes if you think this way is better though.

Reply:
The error is just some kind of print, not exit code.

if `au cypress --run` exit code is 0 (good), in windows powershell, you can use command `echo $?` (have to do right after `au cypress`) to print True.
If exit code is non-zero (means failed), `echo $?` will print False.

For example, CI (like travis-ci or circleci) will just check the exit code to determine whether a command is successful.